files/usr/share/cinnamon/applets/recent@cinnamon.org/applet.js

```diff
@@ -69,14 +70,26 @@ MyApplet.prototype = {
             this._recentButtons = [];
             this._display();
 
-            this.recent_id = this.RecentManager.connect('changed', Lang.bind(this, this._refreshRecents));
-            this.settings_id = this.privacy_settings.connect("changed::" + REMEMBER_RECENT_KEY, Lang.bind(this, this._refreshRecents));
+            this.recent_id = this.RecentManager.connect('changed', Lang.bind(this, this._display));
+            this.settings_id = this.privacy_settings.connect("changed::" + REMEMBER_RECENT_KEY, Lang.bind(this, this._display));
+
+            global.settings.connect('changed::' + PANEL_EDIT_MODE_KEY, Lang.bind(this, this._on_panel_edit_mode_changed));
         }
         catch (e) {
             global.logError(e);
         }
     },
 
+    _on_panel_edit_mode_changed: function () {
+        if (global.settings.get_boolean(PANEL_EDIT_MODE_KEY)) {
+            if (!this.actor.visible) {
+                this.actor.show();
+            }
+        } else {
+            this._refreshRecents();
+        }
+    },
+
```

Comment:
Regarding _on_panel_edit_mode_changed() can we do something more like:

```
if (global.settings.get_boolean(PANEL_EDIT_MODE_KEY)) {
    this.actor.show();
} else {
    this.actor.visible = this._recentButtons.length > 0;
}
```

This gives us something tangible to check (length of _recentButtons) which is established at the end of refreshRecents if no recent files exist.  This keeps us from needing to run _refreshRecents as well.